draft-ietf-httpbis-connect-tcp.md

```diff
@@ -145,11 +145,13 @@ Authentication to a templated TCP proxy normally uses ordinary HTTP authenticati
 
 Clients SHOULD assume that all proxy resources generated by a single template share a protection space (i.e., a realm) ({{?RFC9110, Section 11.5}}).  For many authentication schemes, this will allow the client to avoid waiting for a "401 (Unauthorized)" response before each new connection through the proxy.
 
-## Relationship to the Capsule Protocol
+### Use of the Capsule Protocol {#capsule-protocol}
 
-Unlike the datagram-oriented templated HTTP proxying specifications {{CONNECT-UDP}}{{CONNECT-IP}}, this specification does not make use of the Capsule Protocol {{?RFC9297}}.  A future specification could define a procedure for performing TCP proxying using the Capsule Protocol, but no such procedure is defined here.
+In this specification, use of the Capsule Protocol {{?RFC9297}} is OPTIONAL.  Clients MAY request use of the Capsule Protocol by including a "Capsule-Protocol: ?1" header field in the request.
 
-When implementing this specification, clients and servers MUST NOT send a "Capsule-Protocol: ?1" header field.
+Server support for the Capsule Protocol is also OPTIONAL.  If the request includes "Capsule-Protocol: ?1", and the server does not support the Capsule Protocol, the server MUST respond with a 4xx (Client Error) status and a "Capsule-Protocol: ?0" response header field, and MUST discard any data received on this request stream.  Upon receiving such a response, the client MUST disable use of the Capsule Protocol with this URI Template for the remainder of the session and retry the request.
```

Comment:
You can't infer what end-to-end extensions a server supports from one request, because different requests can be routed to different backends even if they are sent over the same client-intermediary connection. For this to work I think we need to mandate server suppport.

Reply:
That's fine in this case.  The client uses the capsule protocol until it encounters a request where that is rejected.  Then, for the rest of the "session" it sticks to baseline.  Once your server fleet is finished upgrading to capsule support, clients will use it reliably (if they support it).

If you want, we could soften the notion of a "session" to indicate that clients can retry after a reasonable amount of time.

Reply:
"MUST disable for the rest of the session" is a bit strong.  "MUST retry without the Capsule Protocol and MAY disable for all future requests" is all that's actually needed. Let implementations decide how aggressive they want to be about backing off in a heterogeneous environment.

Reply:
@MikeBishop OK, I've made that change.

Reply:
The current text still uses the word "session" without defining what that means. But at this point we're in the weeds, I'm ok with merging this and then figuring out the details in separate smaller issues/PRs